relay/nakama/custom_auth.go

```diff
@@ -0,0 +1,98 @@
+package main
+
+import (
+	"context"
+	"database/sql"
+	"encoding/json"
+	"errors"
+
+	"github.com/heroiclabs/nakama-common/api"
+	"github.com/heroiclabs/nakama-common/runtime"
+	"github.com/rotisserie/eris"
+	"google.golang.org/grpc/codes"
+
+	"pkg.world.dev/world-engine/relay/nakama/siwe"
+	"pkg.world.dev/world-engine/relay/nakama/utils"
+)
+
+const (
+	signInWithEthereumType = "siwe"
+)
+
+var (
+	ErrBadCustomAuthType = errors.New("bad custom auth type")
+)
+
+func initCustomAuthentication(initializer runtime.Initializer) error {
+	if err := initializer.RegisterBeforeAuthenticateCustom(handleCustomAuthentication); err != nil {
+		return eris.Wrap(err, "failed to init siwe authentication")
+	}
+	return nil
+}
+
+func handleCustomAuthentication(
+	ctx context.Context,
+	logger runtime.Logger,
+	_ *sql.DB,
+	nk runtime.NakamaModule,
+	in *api.AuthenticateCustomRequest) (*api.AuthenticateCustomRequest, error) {
+	authType := in.GetAccount().GetVars()["type"]
```

Comment:
it's probably cleaner to have this be a utils function that also validates the authType and return an error if it's not a valid auth type.

Reply:
I'm not sure what you mean by cleaner here. Do you think the chaining of GetAccount().GetVars() isn't clear? 

Regarding validation, the validation code is essentially going to be a duplicate of the code below that identifies and calls the correct handler. The handler identification code will still need to end with a ErrBadCustomAuthType.